[WIP] Azure Devops

cam-barts opened the pull request:
## Azure Devops Integration 
This integrates [Azure Devops](https://azure.microsoft.com/en-us/services/devops/) workitems into Bugwarrior.

So far the service is functional, but there's still some ironing out to do. 

### The general workflow:
- Get workitem ID's from a [wiql](https://docs.microsoft.com/en-us/azure/devops/boards/queries/wiql-syntax?view=azure-devops) query -> list(int)
- For workitem in query, create a taskwarrior issue

### The default query:
```sql
SELECT [System.Id] FROM workitems WHERE [System.AssignedTo] = @me
```
There is a config item where you can add filters to the query. 
For Example: 
```yaml
 ado.wiql_filter = [System.State] not in ('Done', 'Complete') AND [System.WorkItemType] in ('Task', 'Bug')
```

### Framework mods:
- ADO gives options for many work item types that aren't in the standard `cls_markup`
- I use pipenv, so I ignored the Pipfile and Pipfile.lock


### Known Issues:
- With the default query, at least with my actual work, is terribly slow, as it get all previous closed tickets every pull
   - Pulling 185 items takes around a minute to a minute 15 seconds
- It has no logic to get multiple projects/orgs with the same key, so each project/org pair would need a new config target


### Work Still to Be Done
- [ ] Tests
    - [ ] I need to learn nose
- [ ] Documentation
    - [ ] I've not written any docs in restructured text, so I'll have to figure that out too

Fongshway committed:
gmail: replace deprecated oauth2client package with google-auth-oauthlib

Fongshway committed:
Add gmail as supported service in README

Commit:
Add missing codeblock marker

Fongshway committed:
Add TestGmailService

Fongshway committed:
Add test for expired credentials

Fongshway committed:
Refactor test for expired credentials

Fongshway committed:
fixup

cam-barts committed:
Initial Functioning Azure Devops Service

cam-barts committed:
Initial Functioning Azure Devops Service

cam-barts committed:
Ignore pipenv collateral

cam-barts committed:
Ignore pipenv collateral

cam-barts committed:
lowercase the task type in build description

cam-barts committed:
lowercase the task type in build description

ryneeverett commented on .gitignore:
```diff
@@ -18,3 +18,4 @@ man/*
 *.idea
 .tox/*
 *.bak
+Pipfile*
```
FYI, there's been discussion in #743 of adding Pipfile to the git repository as a tracked file.

ryneeverett commented on bugwarrior/services/__init__.py:
```diff
@@ -368,6 +368,8 @@ def build_default_description(
             'chore': 'Chore',
         }
+        if cls not in cls_markup.keys():
+            cls_markup[cls] = cls.lower()
```
It's not clear to me why this is necessary since `task` and `bug` are both keys in `cls_markup`.

cam-barts replied:
Task and Bug aren't the only workitem types that can be assigned to users in Azure Devops. It supports different methodologies, which call tracked tasks different things that aren't included in the dict, like `product backlog item` and `impediment` for scrum process and `change request` and `risk` for CMMI.

ryneeverett replied:
In the past we've added items to `cls_markup` for each individual type, even if it's only used by a single service. (See #666 in which we added 5 items for one service.) It would probably be better to move this logic down into the individual services but I don't think we'll want to accept open-ended types like this.

cam-barts replied:
Perhaps I don't understand this portion of code like I thought I did. I was under the impression that all this provides is a work item type in the default description, and gives it some shorthand or removed it in special cases. I'd be fine doing it either way, and even trying to abstract this into the config somehow, but I don't understand why open ended types would be an issue.

ryneeverett replied:
You're right. In general we prefer that bugwarrior breaks when our API assumptions are violated but in this case there don't appear to be consequences for arbitrary classes.

However, for consistency could we make this  `cls.title()`? You could then remove the last 5 items in `cls_markup` which would be effectively no-op.

Also, I think it would be more straightforward to do `cls_markup.get(cls, cls.title())` in the returned string below than to add it to the dict like this.

pierreprinetti committed:
bugzilla: Treat api_key as a password

This patch enables using a vault or `@oracle` to retrieve a Bugzilla api
key.

Example:

```toml
bugzilla.api_key = @oracle:eval:pass bugzilla/bugwarrior
```

StephanMeijer committed:
#750 document JIRA API tokens

StephanMeijer committed:
#752 Fix JIRA email escaping

jspricke committed:
python-jira parses sprints now

For backward compatibility the old parsing method is still in place.

jspricke committed:
[Jira] support lowercase closed in sprints

ryneeverett committed:
mock (third-party backport) -> unittest.mock

We no longer support any version of python for which this backport is
necessary.

ryneeverett committed:
Bugzilla: Serialize dates in BugzillaService.

Conform to the dominant pattern of deserializing dates in
Issue.to_taskwarrior. Resolve #740.

srl295 committed:
Jira: use paginated query to get ALL results

i haven't read the docs, but from reading the code it seems that `None` is the right thing to pass.  `-1` seems to
get 100 results, while omitting the parameter returns the default of `100`.

Fixes: #764

dthadi3 committed:
Added power support  (#763)

* Update .travis.yml

* Update .travis.yml

Added power architecture...

* Update .travis.yml

given permissions for cache/pip/wheels on ppc64le

ryneeverett committed:
bugzilla: Implement common filters.

This adds support for the common configuration options
`only_if_assigned` and `also_unassigned`, which should be supported by
all services.

Resolve #243.

ryneeverett committed:
Update travis image to newer ubuntu.

This should fix the build on power architecture with a newer version of
openssl.

StephanMeijer committed:
#769 ability to reopen tasks

PtitCaius committed:
Add replace_tags option (#734)

* Adding replace_tags options, allowing user to delete the local tags
before adding the remotes one.

As static_fields, static_tags allow the user to specify some tags to
keep from the local ones.

* Add documentation

djmitche committed:
Add `github.body_length` configuration

This controls the maximum length of the githubbody UDA, and can be used
to avoid syncing the body at all (by setting it to 0).

djmitche committed:
Be more specific in matching configparser errors in config parser

In Python 3, ConfigParser is more exacting about the syntax it allows,
and in particular it disallows bare `%` characters.  This causes an
error for URLs containing such characters, but the bare `except` just
treated this error as a missing configuration and did not show it to the
user.

ryneeverett committed:
bugzilla: make authentication optional

The service works without an account.

ryneeverett committed:
Test reopening tasks.

This tests the behavior described in #769 and implemented in #779.

ryneeverett committed:
1.8.0

ryneeverett committed:
Fix duplicate completed task exception.

This implementation aims to be a more performant alternative to #796 by
avoiding an extra taskwarrior query on each loop.

nblock committed:
Add support for Kanboard service (fixes: #223) (#794)

* Add support for Kanboard service (fixes: #223)

* Add kanboard service and tests
* Add documentation and update README

* Add kanboard to travis

* Mention the use of API tokens instead of a password

* Fix UDA types

* Also use my_api_token in the example configuration

ryneeverett committed:
nose -> pytest

https://nose.readthedocs.io/en/latest#note-to-users

ryneeverett committed:
Test bugwarrior-pull command.

BamBalaam committed:
Add missing UDA Fields in the Docs

silvio commented:
Can I ask what the status of this PR is? This looks like I could make good use of this? Is it still being worked on, @cam-barts ?

cam-barts commented:
Hiya @silvio, I had to put this down for awhile to work on other things, but I can put some effort into it by next week. I need to learn nose to get the cov back up to snuff and I need to address some of the stuff @ryneeverett mentioned, but the functionality is there on my fork and I've been using it for months. Let me know if that works for you.

ryneeverett commented:
> I need to learn nose to get the cov back up to snuff

FYI we switched from nose to pytest since you originally posted this PR.

cam-barts committed:
Merge branch 'azure_devops_service' of https://github.com/cam-barts/bugwarrior into azure_devops_service

cam-barts commented:
It seems I've screwed up my git stuff. Working on cleaning up and then getting some tests in.
